File: javaQuarkusShip/src/test/java/pl/battleships/javaquarkusship/service/GameServiceTest.java
package pl.battleships.javaquarkusship.service;

import io.quarkus.test.junit.QuarkusTest;
import io.quarkus.test.junit.mockito.InjectMock;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.mockito.Mockito;
import pl.battleships.javaquarkusship.api.dto.GameDto;
import pl.battleships.javaquarkusship.api.dto.PositionDto;
import pl.battleships.core.api.BattleshipGame;
import pl.battleships.core.api.HistoryProvider;
import pl.battleships.core.exception.DuplicatedGameException;
import pl.battleships.core.model.Board;
import pl.battleships.core.model.Position;
import pl.battleships.core.model.Ship;

import javax.inject.Inject;
import java.util.List;

@QuarkusTest
class GameServiceTest {

    @InjectMock
    BattleshipGame battleshipGame;

    @InjectMock
    HistoryProvider historyProvider;

    @Inject
    GameServiceImpl gameService;

    @DisplayName("check positive scenario of joining to the game")
    @Test
    void checkJoinTheGame() {
        GameDto game = new GameDto().id("x").size(10);
        Mockito.when(battleshipGame.start(Mockito.any(), Mockito.anyInt(), Mockito.anyBoolean())).thenReturn(
                Board.builder().ships(
                        List.of(
                                Ship.builder().type(1).destroyed(true).location(
                                        List.of(Position.builder().x(1).y(1).hit(true).build())).build(),
                                Ship.builder().type(2).destroyed(false).location(
                                        List.of(Position.builder().x(4).y(4).build())).build())
                ).build()
        );
        gameService.joinTheGame(game);
    }

    @DisplayName("check proper handling of duplicates")
    @Test
    void checkJoinForDuplicates() {
        Mockito.when(battleshipGame.start(Mockito.any(), Mockito.anyInt(), Mockito.anyBoolean())).thenThrow(new DuplicatedGameException());
        //do not catch
        Assertions.assertThrows(DuplicatedGameException.class, () -> gameService.joinTheGame(new GameDto().id("x").size(10)));
    }

    @DisplayName("check getting all shots")
    @Test
    void checkGetAllShots() {
        Mockito.when(historyProvider.getAllShots(Mockito.any())).thenReturn(
                List.of(
                        Position.builder().x(1).y(1).hit(true).build(),
                        Position.builder().x(1).y(2).hit(false).build(),
                        Position.builder().x(1).y(3).hit(false).build()
                )
        );
        List<PositionDto> allShots = gameService.getAllShots("x");
        Assertions.assertEquals(3, allShots.size());
        Assertions.assertEquals(1L, allShots.stream().filter(dto -> Boolean.TRUE.equals(dto.getHit())).count());
    }

}

File: kotlinQuarkusShip/src/test/kotlin/pl/battleships/kotlinquarkusship/service/GameServiceTest.kt
package pl.battleships.kotlinquarkusship.service

import io.mockk.every
import io.mockk.slot
import io.mockk.verify
import io.quarkiverse.test.junit.mockk.InjectMock
import io.quarkus.test.junit.QuarkusTest
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import pl.battleships.api.dto.GameDto
import pl.battleships.api.dto.GameStatusDto
import pl.battleships.api.dto.PositionDto
import pl.battleships.api.dto.ShotStatusDto
import pl.battleships.core.api.BattleshipGame
import pl.battleships.core.api.HistoryProvider
import pl.battleships.core.exception.DuplicatedGameException
import pl.battleships.core.model.*
import java.util.List
import javax.inject.Inject

@QuarkusTest
class GameServiceTest {

    @InjectMock
    lateinit var battleshipGame: BattleshipGame

    @InjectMock
    lateinit var historyProvider: HistoryProvider

    @Inject
    lateinit var gameService: GameServiceImpl

    @Test
    fun `check getting all shots`() {
        every { historyProvider.getAllShots(any()) } returns listOf(
            Position.builder().x(1).y(1).hit(false).build(),
        ) andThen listOf(
            Position.builder().x(1).y(1).hit(false).build(),
            Position.builder().x(2).y(2).hit(true).build()
        )

        val shots = gameService.getAllShots("x")
        assertEquals(1, shots.size)

        val shots2 = gameService.getAllShots("y")
        assertEquals(2, shots2.size)


        verify(exactly = 2) {
            historyProvider.getAllShots(any())
        }
    }

    @Test
    fun `check proper handling of duplicates`() {
        val slot = slot<String>()
        every { battleshipGame.start(capture(slot), any(), any()) } throws DuplicatedGameException()

        assertThrows<DuplicatedGameException> {
            gameService.joinTheGame(GameDto("x", 10, true))
        }

        assertEquals("x", slot.captured)
    }

    @Test
    fun `check positive scenario of joining to the game`() {
        every { battleshipGame.start(any(), any(), any()) } returns Board.builder().ships(
            List.of(
                Ship.builder().type(1).destroyed(true).location(
                    List.of(Position.builder().x(1).y(1).hit(true).build())
                ).build(),
                Ship.builder().type(2).destroyed(false).location(
                    List.of(Position.builder().x(4).y(4).build())
                ).build()
            )
        ).build()

        gameService.joinTheGame(GameDto("x", 10, true))

        verify {
            battleshipGame.start(any(), any(), any())
        }
    }

    @Test
    fun `check game status`() {
        every { battleshipGame.getGameStatus(any()) } returns GameStatus.RUNNING

        val gameStatus = gameService.getGameStatus("x")
        assertEquals(GameStatusDto.RUNNING, gameStatus)
    }

    @Test
    fun `check opponent shot`() {
        every { battleshipGame.opponentShot(any(), any()) } returns ShotResult.HIT

        val result = gameService.opponentShot("x", PositionDto(1, 2))
        assertEquals(ShotStatusDto.HIT, result)
    }
}